crates/re_build_build_info/src/lib.rs

```diff
@@ -0,0 +1,96 @@
+//! This crate is to be used from `build.rs` build scripts.
+//!
+//! Use this crate together with the `re_build_info` crate.
+
+use anyhow::Context as _;
+
+use std::process::Command;
+
+// Situations to consider
+// ----------------------
+//
+// # Using the published crate
+//
+// The published crate carries its version around, which in turns gives us the git tag, which makes
+// the commit hash irrelevant.
+// We still need to compute _something_ so that we can actually build, but that value will be
+// ignored when the crate is built by the end user anyhow.
+//
+// # Working directly within the workspace
+//
+// When working within the workspace, we can simply try and call `git` and we're done.
+//
+// # Using an unpublished crate (e.g. `path = "..."` or `git = "..."` or `[patch.crates-io]`)
+//
+// In these cases we may or may not have access to the workspace (e.g. a `path = ...` import likely
+// will, while a crate patch won't).
+//
+// This is not an issue however, as we can simply try and see what we get.
+// If we manage to compute a commit hash, great, otherwise we still have the crate version to
+// fallback on.
+
+/// Call from the `build.rs` file of any crate you want to generate build info for.
+pub fn export_env_vars() {
+    // target triple
+    println!(
+        "cargo:rustc-env=RE_BUILD_TARGET_TRIPLE={}",
+        std::env::var("TARGET").unwrap()
+    );
+
+    if let Ok(git_hash) = git_hash() {
+        println!("cargo:rustc-env=RE_BUILD_GIT_HASH={git_hash}");
+        for path in glob::glob("../../.git/refs/heads/**").unwrap() {
+            println!("cargo:rerun-if-changed={}", path.unwrap().to_string_lossy());
+        }
+    } else {
+        // NOTE: In 99% of cases, if `git_hash` failed it's because we're not in a git repository
+        // to begin with, which happens because we've imported the published crate from crates.io.
+        //
+        // Example of unlikely cases where the above does not hold:
+        // - `git` is not installed
+        // - the user downloaded rerun as a tarball and then imported via a `path = ...` import
+        // - others?
+        println!("cargo:rustc-env=RE_BUILD_GIT_HASH=");
+    }
+
+    let git_branch = git_branch().unwrap_or_default();
+    println!("cargo:rustc-env=RE_BUILD_GIT_BRANCH={git_branch}");
+
+    let is_git_clean = is_git_clean().unwrap_or_default();
+    println!("cargo:rustc-env=RE_BUILD_GIT_IS_CLEAN={is_git_clean}");
```

Comment:
I feel like we shouldn't even try this? the build.rs would technically need to track for _any_ change in the repo for this value to be correct! (and thankfully this doesn't so far).
I mean right now doing a change that should make this one flip to dirty won't actually work or am I missing something?

Reply:
You are right, this doesn't work:

```
❯ cargo r -p rerun -- --version
rerun 0.2.0 aarch64-apple-darwin emilk/build-info 193dd94, built 2023-02-25T21:30:24Z

❯ echo "// foo" >> crates/re_ui/src/lib.rs

❯ cargo r -p rerun -- --version           
   Compiling re_ui v0.2.0 (/Users/emilk/code/rerun/rerun/crates/re_ui)
   Compiling re_viewer v0.2.0 (/Users/emilk/code/rerun/rerun/crates/re_viewer)
   Compiling re_sdk v0.2.0 (/Users/emilk/code/rerun/rerun/crates/re_sdk)
   Compiling rerun v0.2.0 (/Users/emilk/code/rerun/rerun/crates/rerun)
    Finished dev [optimized + debuginfo] target(s) in 22.02s
     Running `target/debug/rerun --version`
rerun 0.2.0 aarch64-apple-darwin emilk/build-info 193dd94, built 2023-02-25T21:30:24Z
```

@teh-cmc wrote the original code, maybe he has a better idea?

I'd be happy to scrap it though.

Reply:
I agree that this doesn't seem feasible to get right, so I think we should just scrap `-dirty` from our analytics (for which this feature was originally added).

Reply:
There is a convoluted way to make the dirtiness check work even if there are no local commits (see https://github.com/rerun-io/rerun/issues/925: basically you have one build.rs file that runs all the time and just checks for dirtiness, and if it finds any, it touches a file that re_analytics's build.rs is watching for).

I do think we should either go all the way or scrap it entirely; other than that I have no preference.

Reply:
I removed the `-dirty` thing for now. It is not worth the complexity and flakyness.